freqtrade/freqai/transforms/quantile_transform.py

```diff
@@ -0,0 +1,28 @@
+from sklearn.preprocessing import QuantileTransformer
+
+
+class FreqaiQuantileTransformer(QuantileTransformer):
+    """
+    A subclass of the SKLearn Quantile that ensures fit, transform, fit_transform and
+    inverse_transform all take the full set of params X, y, sample_weight required to
+    benefit from the DataSieve features.
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def fit_transform(self, X, y=None, sample_weight=None, feature_list=None, **kwargs):
+        super().fit(X)
```

Comment:
It is not the case, so SKLearn's `fit()` is not the same as the `self.fit()`. In this case it is because we want to carry these 3 extra variables through our pipeline. 

If you have a recommended way to change this I am all ears - but this seemed to be the most effective approach to keep SKLearn around and keep a familiar API, while introducing the possibility of making actions on the other variables.

Reply:
mmh, that's not exactly what i meant, maybe i should've written `self.fit(X, y, sample_weight, feature_list, **kwargs)` so it's clear which method i think should be called.

in my understanding, `fit_transform()` is shorthand to `.fit() & .transform()` (conceptually - which is somewhat confirmed by the calls to `super().fit()` and `super().transform()` in this method).

As such, i could inherit from this class, implement logic _only_ in `.fit()` (the rest of the methods is already implemented, as this is not an abstract interface).
I'd then expect my `fit_transform()` method to call my newly created logic in `.fit()` (obviously the same applies to the `.transform()` method below.
by calling `super().fit()` here - changes to the implementation of `.fit()` will be ignored when calling `.fit_transfrom()`.

Reply:
Yeah, you are right. In fact, that is how we already handle the more interesting transforms, like the `SVMOutlierExtractor` (the conversion of our `use_svm_to_remove_outliers()`)  [here](https://github.com/emergentmethods/datasieve/blob/main/datasieve/transforms/svm_outlier_extractor.py#L19).

 This FreqaiQuantileTransform is a very simple example of how to get an SKLearn element into the FreqAI pipeline.  Without a doubt, putting `self.fit(); self.transform()` into `fit_transform()` would be superior for this example (Ill swap it over) as it would allow users to continue inheriting from this example if they wished. Which is why we do exactly that with `SVMOuterlierExtractor`. However, we agree that running this simple code with `self.fit()` or `super.fit()` yields the same results. 

The simplicity of this example doesnt help convey they general idea of carrying `X, y, sample_weights, feature_list` around together. So I would implore you to check out the `SVMOutlierExtractor` linked above - it would help illustrate the ultimate needs/goals.

Reply:
Yeah, that's what i mean.

obviously for **this** implementation, it's irrelevant, but further inheriting from this (only implementing what's needed) might have this suddenly buggy without good reason.

---

Another interesting point the link to the `svm_outlier_extractor` had me think about is that `.fit()` actually returns a value (i assume X?) - which is however not used.
this will have no impact as long as the original `X, y, ...` are mutated - but if for some reason, a new list is assigned / returned (for whatever reason) instead of mutating the input, then that new list would be ignored, effectively / potentially causing faulty/wrong behavior.
Again, most likely no impact in this particular instance (accidentally), but not really suitable and potentially problematic for a functional interface.

Effectively, this might mean that this line should potentially be something around the below (obviously, not just here)

```
X, y, sample_weight, feature_list = self.fit(X, y, sample_weight, feature_list, **kwargs)
# ...
```

Reply:
It depends entirely on each step of the pipeline. If the user thinks that their `fit()` should modify `X`, then they would return and pass it along. For SVMOutlierExtractor `fit()` simply fits the model and does not `transform()` anything. This is the typical case in SKLearn as well, `fit()` does not mean `transform()`. If you want to also `transform`,then you call `fit_transform()`. 

But it doesnt stop users from doing anything they want inside `fit()` if they think that `fit()` should be doing more to prepare for the subsequent `fit()` functions in other steps. 

For example, PCA really doesnt work unless it is `fit_transform` or `transform`. This is because the PCA step changes the dimensionality of the data, so subsequent steps *depend* on this transformation to be properly fit. Else they are useless. 

Building pipelines gets tricky when incorporating a wide range of methods and when trying to accommodate a variety of input/output needs.